File: dayhye/Programmers/연습문제/lv2/거리두기 확인하기.py
def check_place(arr, x, y):
    dx = [-1, 0, 1, 0, -2, 0, 2, 0, -1, 1, 1, -1]
    dy = [0, 1, 0, -1, 0, 2, 0, -2, 1, 1, -1, -1]

    temp = 1
    for i in range(12):
        nx = x + dx[i]
        ny = y + dy[i]

        if 0 <= nx < 5 and 0 <= ny < 5:
            if i < 4:
                if check_person(arr, nx, ny):
                    temp = 0
                    break
            elif i < 8:
                if check_person(arr, nx, ny):
                    if i == 4:
                        if not arr[x - 1][y] == 'X':
                            temp = 0
                            break
                    elif i == 5:
                        if not arr[x][y + 1] == 'X':
                            temp = 0
                            break
                    elif i == 6:
                        if not arr[x + 1][y] == 'X':
                            temp = 0
                            break
                    else:
                        if not arr[x][y - 1] == 'X':
                            temp = 0
                            break
            else:
                if check_person(arr, nx, ny):
                    if i == 8:
                        if not arr[x - 1][y] == 'X' or not arr[x][y + 1] == 'X':
                            temp = 0
                            break

                    elif i == 9:
                        if not arr[x][y + 1] == 'X' or not arr[x + 1][y] == 'X':
                            temp = 0
                            break

                    elif i == 10:
                        if not arr[x][y - 1] == 'X' or not arr[x + 1][y] == 'X':
                            temp = 0
                            break

                    elif i == 11:
                        if not arr[x - 1][y] == 'X' or not arr[x][y - 1] == 'X':
                            temp = 0
                            break

    return temp


def check_person(arr, i, j):
    if arr[i][j] == 'P':
        return True
    return False


def solution(places):
    answer = []

    for place in places:

        arr = []

        for line in place:
            temp = list(line)
            arr.append(temp)

        result = 1

        for i in range(5):
            for j in range(5):
                if arr[i][j] == 'P':

                    result = check_place(arr, i, j)
                    if result == 0:
                        break

            if result == 0:
                break

        answer.append(result)

    return answer

File: eunnyeong/거리두기 확인하기.java
class Solution {
    int[] dx = {-1, 1, 0, 0, -1, -1, 1, 1}, dy = {0, 0, -1, 1, -1, 1, -1, 1};
    int[][] select = {{-1, -1}, {-1, -1}, {-1, -1}, {-1, -1}, {0, 2}, {0, 3}, {1, 2}, {1, 3}};
    
    public int[] solution(String[][] places) {
        int[] answer = new int[5];
        
        for (int i = 0; i < 5; i++) {
            char[][] c = new char[5][5];
            
            for (int j = 0; j < 5; j++)
                c[j] = places[i][j].toCharArray();
    
            answer[i] = check(c);
        }
        
        return answer;
    }
    
    public int check(char[][] c) {
        for (int i = 0; i < 5; i++) {
            for (int j = 0; j < 5; j++) {
                if (c[i][j] == 'P') {
                    for (int l = 1; l <= 2; l++) {
                        for (int k = 0; k < 4; k++) {
                            int nx = i + dx[k] * l, ny = j + dy[k] * l;
                            if (range(nx, ny) && c[nx][ny] == 'P' && c[nx - dx[k]][ny - dy[k]] != 'X') {
                                return 0;
                            }
                        }
                    }
                    
                    for (int k = 4; k < 8; k++) {
                        int nx = i + dx[k], ny = j + dy[k];
                        if (range(nx, ny) && c[nx][ny] == 'P') {
                            int f = 0;
                            for (int l = 0; l < 2; l++) {
                                int tx = i + dx[select[k][l]], ty = j + dy[select[k][l]];
                                if (range(tx, ty) && c[tx][ty] == 'X')
                                    f++;
                            }
                            if (f != 2)
                                return 0;
                        }
                    }
                }
            }
        }
        
        return 1;
    }
    
    public boolean range(int nx, int ny) {
        return nx >= 0 && nx < 5 && ny >= 0 && ny < 5;
    }
}
